capsules/src/rf233.rs

```diff
@@ -0,0 +1,885 @@
+                                            (self.addr.get() >> 8) as u8,
+                                            InternalState::CONFIG_SHORT1_SET);
+            }
+            InternalState::CONFIG_SHORT1_SET => {
+                self.state_transition_write(RF233Register::PAN_ID_0,
+                                            (self.pan.get() & 0xff) as u8,
+                                            InternalState::CONFIG_PAN0_SET);
+            }
+            InternalState::CONFIG_PAN0_SET => {
+                self.state_transition_write(RF233Register::PAN_ID_1,
+                                            (self.pan.get() >> 8) as u8,
+                                            InternalState::READY);
+            }
+            InternalState::UNKNOWN => {}
+        }
+    }
+}
+
+impl<'a, S: spi::SpiMasterDevice + 'a> gpio::Client for RF233<'a, S> {
+    fn fired(&self, identifier: usize) {
+        if identifier == INTERRUPT_ID {
+            self.handle_interrupt();
+        }
+    }
+}
+
+impl<'a, S: spi::SpiMasterDevice + 'a> RF233<'a, S> {
+    pub fn new(spi: &'a S,
+               reset: &'a gpio::Pin,
+               sleep: &'a gpio::Pin,
+               irq: &'a gpio::Pin,
+               ctl: &'a gpio::PinCtl)
+               -> RF233<'a, S> {
+        RF233 {
+            spi: spi,
+            reset_pin: reset,
+            sleep_pin: sleep,
+            irq_pin: irq,
+            irq_ctl: ctl,
+            radio_on: Cell::new(false),
+            transmitting: Cell::new(false),
+            receiving: Cell::new(false),
+            spi_busy: Cell::new(false),
+            state: Cell::new(InternalState::START),
+            interrupt_handling: Cell::new(false),
+            interrupt_pending: Cell::new(false),
+            tx_buf: TakeCell::empty(),
+            rx_buf: TakeCell::empty(),
+            tx_len: Cell::new(0),
+            tx_client: Cell::new(None),
+            rx_client: Cell::new(None),
+            addr: Cell::new(0),
+            pan: Cell::new(0),
+            seq: Cell::new(0),
+            spi_rx: TakeCell::empty(),
+            spi_tx: TakeCell::empty(),
+            spi_buf: TakeCell::empty(),
+        }
+    }
+
+    fn handle_interrupt(&self) {
+        // Because the first thing we do on handling an interrupt is
+        // read the IRQ status, we defer handling the state transition
+        // to the SPI handler
+        if self.spi_busy.get() == false {
+            self.interrupt_handling.set(true);
+            self.register_read(RF233Register::IRQ_STATUS);
+        } else {
+            self.interrupt_pending.set(true);
+        }
+    }
+
+    #[allow(dead_code)]
+    fn register_write(&self, reg: RF233Register, val: u8) -> ReturnCode {
+
+        if (self.spi_busy.get() || self.spi_tx.is_none() || self.spi_rx.is_none()) {
+            return ReturnCode::EBUSY;
+        }
+        let wbuf = self.spi_tx.take().unwrap();
+        let rbuf = self.spi_rx.take().unwrap();
```

Comment:
I'm with @ppannuto on this. We need to draw a hard line a hard-line about `unwrap` in untrusted capsules. A, maybe hacky, alternative, which I believe would work here, is to use `unwrap_or(&mut [])`:

```
let wbuf = self.spi_tx.take().unwrap_or(&mut []);
```

Since it's an empty slice, the reference can be inferred to have a static lifetime, so the types work out. As the capsule writer, you _know_ you'll never actually have to deal with that case because you just checked a second ago that it's non-empty. As an auditor, I know that even if you're logic is wrong, worst case the driver won't work as expected, but it won't panic.